File: android/src/main/java/com/margelo/nitro/svga/ScaleCalculator.kt
package com.margelo.nitro.svga

internal object ScaleCalculator {

  data class Result(val scaleX: Float, val scaleY: Float, val translateX: Float, val translateY: Float)

  fun compute(
    mode: ScaleMode,
    viewWidth: Float,
    viewHeight: Float,
    contentWidth: Float,
    contentHeight: Float
  ): Result {
    if (contentWidth <= 0f || contentHeight <= 0f) return Result(1f, 1f, 0f, 0f)

    if (mode == ScaleMode.FILL) {
      return Result(viewWidth / contentWidth, viewHeight / contentHeight, 0f, 0f)
    }

    val sx = viewWidth / contentWidth
    val sy = viewHeight / contentHeight
    val scale = when (mode) {
      ScaleMode.ASPECTFILL -> maxOf(sx, sy)
      else -> minOf(sx, sy)
    }
    val tx = (viewWidth - contentWidth * scale) / 2f
    val ty = (viewHeight - contentHeight * scale) / 2f
    return Result(scale, scale, tx, ty)
  }
}

File: ios/ScaleCalculator.swift
import CoreGraphics

internal struct ScaleResult {
    let scaleX: CGFloat
    let scaleY: CGFloat
    let translateX: CGFloat
    let translateY: CGFloat

    /// The viewBox → view-space affine: a content point is scaled, then
    /// translated (centering offset). `a.concatenating(b)` applies `a` then
    /// `b`, so this is scale-then-translate. The renderer pre-multiplies each
    /// sprite's frame transform onto this (`frame.transform.concatenating(.)`),
    /// reproducing the old `draw(rect:)` CTM order exactly. Kept here as the
    /// single source of truth so `SvgaPlayerView` and its tests agree.
    var transform: CGAffineTransform {
        return CGAffineTransform(scaleX: scaleX, y: scaleY)
            .concatenating(CGAffineTransform(translationX: translateX, y: translateY))
    }
}

internal enum ScaleCalc {
    static func compute(
        mode: ScaleMode,
        viewWidth: CGFloat,
        viewHeight: CGFloat,
        contentWidth: CGFloat,
        contentHeight: CGFloat
    ) -> ScaleResult {
        if contentWidth <= 0 || contentHeight <= 0 {
            return ScaleResult(scaleX: 1, scaleY: 1, translateX: 0, translateY: 0)
        }
        if mode == .fill {
            return ScaleResult(
                scaleX: viewWidth / contentWidth,
                scaleY: viewHeight / contentHeight,
                translateX: 0,
                translateY: 0
            )
        }
        let sx = viewWidth / contentWidth
        let sy = viewHeight / contentHeight
        let scale: CGFloat
        switch mode {
        case .aspectfill: scale = max(sx, sy)
        default: scale = min(sx, sy)
        }
        let tx = (viewWidth - contentWidth * scale) / 2
        let ty = (viewHeight - contentHeight * scale) / 2
        return ScaleResult(scaleX: scale, scaleY: scale, translateX: tx, translateY: ty)
    }
}
